### PR TITLE
Add option for time smearing and TOF compensation in DDPlanarDigiProcessor

### PR DESCRIPTION
**Pull request by RCLanger**
BEGINRELEASENOTES
- Added time smearing to the DDPlanarDigiProcessor 
- Introduced a time window for digitized hits in the DDPlanarDigiProcessor 
- Introduced TOF compensation for digitized hits in the DDPlanarDigiProcessor 

ENDRELEASENOTES

### CODE REVIEW

**Comment by vvolkl** [comment]
@andresailer could your review this? Thanks!

File: source/Digitisers/src/DDPlanarDigiProcessor.cc [hunk 1]
```diff
@@ -308,10 +348,41 @@ void DDPlanarDigiProcessor::processEvent( LCEvent * evt ) {
...
       }
+
+      //***************************************************************
+      // Smear time of the hit and apply the time window cut if needed
+      //***************************************************************
       
-      //**************************************************************************
-      // Try to smear the hit but ensure the hit is inside the sensitive region
-      //**************************************************************************
+      // Smearing time of the hit
```
Comment (andresailer) [inline source/Digitisers/src/DDPlanarDigiProcessor.cc:356]:
```suggestion
      double hitT = simTHit->getTime();
      // Smearing time of the hit
      if (_resT.size() and _resT[0] > 0.0) {
```
indent and close at the end.

File: source/Digitisers/src/DDPlanarDigiProcessor.cc [hunk 2]
```diff
@@ -308,10 +348,41 @@ void DDPlanarDigiProcessor::processEvent( LCEvent * evt ) {
...
+      float resT = _resT.size() > 1 ? _resT.at(layer) : _resT.at(0); 
+      double tSmear  = resT == 0.0 ? 0.0 : gsl_ran_gaussian( _rng, resT );
+      _h[hT]->Fill( resT == 0.0 ? 0.0 : tSmear / resT );
+      _h[diffT]->Fill( tSmear );
+
+      // Skipping the hit if its time is outside the acceptance time window
+      double hitT = simTHit->getTime() + tSmear;
```
Comment (andresailer) [inline source/Digitisers/src/DDPlanarDigiProcessor.cc:363]:
```suggestion
      hitT += tSmear;
```

File: source/Digitisers/src/DDPlanarDigiProcessor.cc [hunk 3]
```diff
@@ -308,10 +348,41 @@ void DDPlanarDigiProcessor::processEvent( LCEvent * evt ) {
...
+      //***************************************************************
+      // Smear time of the hit and apply the time window cut if needed
+      //***************************************************************
       
-      //**************************************************************************
-      // Try to smear the hit but ensure the hit is inside the sensitive region
-      //**************************************************************************
+      // Smearing time of the hit
+      float resT = _resT.size() > 1 ? _resT.at(layer) : _resT.at(0); 
+      double tSmear  = resT == 0.0 ? 0.0 : gsl_ran_gaussian( _rng, resT );
```
Comment (andresailer) [inline source/Digitisers/src/DDPlanarDigiProcessor.cc:358]:
```suggestion
      double tSmear  = resT > 0.0 : gsl_ran_gaussian( _rng, resT ) ? 0.0;
```
I guess it is possible that one would like to disable smearing for some layers, so we need to keep testing

File: source/Digitisers/src/DDPlanarDigiProcessor.cc [hunk 4]
```diff
@@ -308,10 +348,41 @@ void DDPlanarDigiProcessor::processEvent( LCEvent * evt ) {
...
+      _h[hT]->Fill( resT == 0.0 ? 0.0 : tSmear / resT );
```
Comment (andresailer) [inline source/Digitisers/src/DDPlanarDigiProcessor.cc:359]:
```suggestion
      _h[hT]->Fill( resT > 0.0 ? tSmear / resT : 0.0);
```

File: source/Digitisers/src/DDPlanarDigiProcessor.cc [hunk 5]
```diff
@@ -308,10 +348,41 @@ void DDPlanarDigiProcessor::processEvent( LCEvent * evt ) {
...
+
+      // Correcting for the propagation time
+      if (_correctTimesForPropagation) {
+        double dt = oldPos.r() / ( TMath::C() / 1e6 );
+        hitT -= dt;
+        streamlog_out(DEBUG3) << "corrected hit at R: " << oldPos.r() << " mm by propagation time: " << dt << " ns to T: " << hitT << " ns" << std::endl;
+      }
+      
+      if (_useTimeWindow && ( hitT < timeWindow_min || hitT > timeWindow_max) ) {
+        streamlog_out(DEBUG4) << "hit at T: " << hitT << " smeared to: " << hitT << " is outside the time window: hit dropped"  << std::endl;
```
Comment (andresailer) [inline source/Digitisers/src/DDPlanarDigiProcessor.cc:377]:
```suggestion
        streamlog_out(DEBUG4) << "hit at T: " << simTHit->getTime() << " smeared to: " << hitT << " is outside the time window: hit dropped"  << std::endl;
```

File: source/Digitisers/src/DDPlanarDigiProcessor.cc [hunk 6]
```diff
@@ -308,10 +348,41 @@ void DDPlanarDigiProcessor::processEvent( LCEvent * evt ) {
...
+      double tSmear  = resT == 0.0 ? 0.0 : gsl_ran_gaussian( _rng, resT );
+      _h[hT]->Fill( resT == 0.0 ? 0.0 : tSmear / resT );
+      _h[diffT]->Fill( tSmear );
+
+      // Skipping the hit if its time is outside the acceptance time window
+      double hitT = simTHit->getTime() + tSmear;
+      streamlog_out(DEBUG3) << "smeared hit at T: " << simTHit->getTime() << " ns to T: " << hitT << " ns according to resolution: " << resT << " ns" << std::endl;
+      
+      float timeWindow_min = _timeWindow_min.size() > 1 ? _timeWindow_min.at(layer) : _timeWindow_min.at(0);
+      float timeWindow_max = _timeWindow_max.size() > 1 ? _timeWindow_max.at(layer) : _timeWindow_max.at(0);
```
Comment (andresailer) [inline source/Digitisers/src/DDPlanarDigiProcessor.cc:367]: These lines should be moved to inside the `if(_useTimeWindow)`, and the if for the window selection bet separated out.

File: source/Digitisers/src/DDPlanarDigiProcessor.cc [hunk 7]
```diff
@@ -308,10 +348,41 @@ void DDPlanarDigiProcessor::processEvent( LCEvent * evt ) {
...
+      // Smearing time of the hit
+      float resT = _resT.size() > 1 ? _resT.at(layer) : _resT.at(0); 
+      double tSmear  = resT == 0.0 ? 0.0 : gsl_ran_gaussian( _rng, resT );
+      _h[hT]->Fill( resT == 0.0 ? 0.0 : tSmear / resT );
+      _h[diffT]->Fill( tSmear );
+
+      // Skipping the hit if its time is outside the acceptance time window
+      double hitT = simTHit->getTime() + tSmear;
+      streamlog_out(DEBUG3) << "smeared hit at T: " << simTHit->getTime() << " ns to T: " << hitT << " ns according to resolution: " << resT << " ns" << std::endl;
+
```
Comment (andresailer) [inline source/Digitisers/src/DDPlanarDigiProcessor.cc:365]:
```suggestion
      } // if time smearing enabled
```
As time correction or time window selection might be interesting also without smearing

**Review by andresailer — COMMENTED** [review]
I am feeling a bit queasy about the direct comparison with the floats.
And it should be possible to essentially have no overhead if the time smearing is disabled.
I hope my proposals are coherent, but needs checking...

**Comment by andresailer** [comment]
Please rebase to get the CI builds (let me know if you need some advice on rebasing).

File: source/Digitisers/src/DDPlanarDigiProcessor.cc [hunk 8]
```diff
@@ -308,10 +348,46 @@ void DDPlanarDigiProcessor::processEvent( LCEvent * evt ) {
...
+      
+      // Skipping the hit if its time is outside the acceptance time window
+      if (_useTimeWindow) {
+        float timeWindow_min = _timeWindow_min.size() > 1 ? _timeWindow_min.at(layer) : _timeWindow_min.at(0);
+        float timeWindow_max = _timeWindow_max.size() > 1 ? _timeWindow_max.at(layer) : _timeWindow_max.at(0);
+        if ( hitT < timeWindow_min || hitT > timeWindow_max ) {
+        streamlog_out(DEBUG4) << "hit at T: " << simTHit->getTime() << " smeared to: " << hitT << " is outside the time window: hit dropped"  << std::endl;
+        ++nDismissedHits;
+        }
+        continue;
```
Comment (andresailer) [inline source/Digitisers/src/DDPlanarDigiProcessor.cc:384]: The `continue` has to be inside the if before. (and the statements in the if need to be indented)

File: source/Digitisers/src/DDPlanarDigiProcessor.cc [hunk 9]
```diff
@@ -308,10 +348,46 @@ void DDPlanarDigiProcessor::processEvent( LCEvent * evt ) {
...
+      
+      // Smearing time of the hit
+      if (_resT.size() and _resT[0] > 0.0) {
+      float resT = _resT.size() > 1 ? _resT.at(layer) : _resT.at(0); 
+      double tSmear  = resT > 0.0 ? gsl_ran_gaussian( _rng, resT ) : 0.0;
+      _h[hT]->Fill( resT > 0.0 ? tSmear / resT : 0.0 );
+      _h[diffT]->Fill( tSmear );
+
+      hitT += tSmear;
+      streamlog_out(DEBUG3) << "smeared hit at T: " << simTHit->getTime() << " ns to T: " << hitT << " ns according to resolution: " << resT << " ns" << std::endl;
```
Comment (andresailer) [inline source/Digitisers/src/DDPlanarDigiProcessor.cc:366]: Please indent all these statements

**Review by andresailer — APPROVED** [review]

**Comment by andresailer** [comment]
replaced by #53

**Pull request closed without merging**